### PR TITLE
[CI/Build] Add TP test for vision models

### PR DESCRIPTION
**Pull request by DarkLight1337**
Follow-up to #5880

~~**Note:** This test is not run in CI as it requires multiple GPUs. You can still try it locally though.~~

cc @xwjiang2010 @ywang96

### CODE REVIEW

**Comment by DarkLight1337** [comment]
Currently getting this error:

```
tests/conftest.py:493: in generate_greedy
    outputs = self.generate(prompts, greedy_params, images=images)
tests/conftest.py:451: in generate
    req_outputs = self.model.generate(inputs,
vllm/utils.py:764: in inner
    return fn(*args, **kwargs)
vllm/entrypoints/llm.py:304: in generate
    outputs = self._run_engine(use_tqdm=use_tqdm)
vllm/entrypoints/llm.py:556: in _run_engine
    step_outputs = self.llm_engine.step()
vllm/engine/llm_engine.py:806: in step
    output = self.model_executor.execute_model(
vllm/executor/distributed_gpu_executor.py:76: in execute_model
    return self._driver_execute_model(execute_model_req)
vllm/executor/multiproc_gpu_executor.py:88: in _driver_execute_model
    return self.driver_worker.execute_model(execute_model_req)
vllm/worker/worker_base.py:236: in execute_model
    broadcast_tensor_dict(broadcast_data, src=0)
vllm/distributed/communication_op.py:32: in broadcast_tensor_dict
    return get_tp_group().broadcast_tensor_dict(tensor_dict, src)
vllm/distributed/parallel_state.py:482: in broadcast_tensor_dict
    self.broadcast_object(metadata_list, src=src)
vllm/distributed/parallel_state.py:358: in broadcast_object
    return self.shm_broadcaster.broadcast_object(obj)
vllm/distributed/device_communicators/shm_broadcast.py:266: in broadcast_object
    self.enqueue(obj)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <vllm.distributed.device_communicators.shm_broadcast.ShmRingBufferIO object at 0x7fe688151e10>
obj = [('num_seq_groups', 2), ('blocks_to_swap_in', TensorMetadata(device='cpu', dtype=torch.int64, size=torch.Size([0, 2]))...ch.Size([1186]))), ('input_positions', TensorMetadata(device='cuda', dtype=torch.int64, size=torch.Size([1186]))), ...]

    def enqueue(self, obj):
        assert self._is_writer, "Only writers can enqueue"
        serialized_obj = pickle.dumps(obj, protocol=pickle.HIGHEST_PROTOCOL)
        if len(serialized_obj) > self.buffer.max_chunk_bytes:
>           raise RuntimeError(
                f"{len(serialized_obj)=} larger than the allowed value "
                f"{self.buffer.max_chunk_bytes},"
                "Please increase the max_chunk_bytes parameter.")
E           RuntimeError: len(serialized_obj)=1356171 larger than the allowed value 1048576,Please increase the max_chunk_bytes parameter.

vllm/distributed/device_communicators/shm_broadcast.py:248: RuntimeError
```

**Comment by DarkLight1337** [comment]
It seems that the program hangs when two tests with `tensor_parallel_size>1` are run after each other. @youkaichao can you help look into this?

**Comment by DarkLight1337** [comment]
> what is the data you try to send? why is it so large?

It's sending the multimodal data (image).

> for distributed related tests, we need to put it under `tests/distributed` folder, and run it one by one.

Alright, thanks for the info!

**Comment by youkaichao** [comment]
do we have any size limit for the image input?

**Comment by DarkLight1337** [comment]
> do we have any size limit for the image input?

Not really. The size depends on the model and type of data. From my understanding, the multi-modal data is send via this:

https://github.com/vllm-project/vllm/blob/3fd02bda51ee7cf07e0375994ac1f34b6d1b981b/vllm/worker/model_runner.py#L73-L82

**Comment by DarkLight1337** [comment]
Based on the results of the tests, it is safe to remove the `.to(device)` calls inside the VLMs.

**Comment by DarkLight1337** [comment]
@youkaichao Please take a look again and see if it is ok. The buffer size is set to 4x compared to before this PR.

File: .buildkite/test-pipeline.yaml [hunk 1]
```diff
@@ -56,6 +57,8 @@ steps:
   - TEST_DIST_MODEL=meta-llama/Llama-2-7b-hf DISTRIBUTED_EXECUTOR_BACKEND=mp pytest -v -s distributed/test_basic_distributed_correctness.py
   - TEST_DIST_MODEL=facebook/opt-125m DISTRIBUTED_EXECUTOR_BACKEND=mp pytest -v -s distributed/test_chunked_prefill_distributed.py
   - TEST_DIST_MODEL=meta-llama/Llama-2-7b-hf DISTRIBUTED_EXECUTOR_BACKEND=mp pytest -v -s distributed/test_chunked_prefill_distributed.py
+  - TEST_DIST_MODEL=llava-hf/llava-1.5-7b-hf DISTRIBUTED_EXECUTOR_BACKEND=mp pytest -v -s distributed/test_multimodal_broadcast.py
+  - TEST_DIST_MODEL=microsoft/Phi-3-vision-128k-instruct DISTRIBUTED_EXECUTOR_BACKEND=mp pytest -v -s distributed/test_multimodal_broadcast.py
```
Comment (youkaichao) [inline .buildkite/test-pipeline.yaml:61]: also add tests with ray backend

**Comment by youkaichao** [comment]
> The buffer size is set to 4x compared to before this PR.

increasing the buffer size looks good to me. CPU memory is cheap.

> Note: This test is not run in CI as it requires multiple GPUs. You can still try it locally though.

what does this mean?

This PR looks good to me in general, and I will hand it over to @ywang96 for the final review regarding vision language models.

**Comment by DarkLight1337** [comment]
>> Note: This test is not run in CI as it requires multiple GPUs. You can still try it locally though.
> what does this mean?

Oops, I forgot to update the OP after moving the tests. They are now run under distributed tests CI.

**Comment by DarkLight1337** [comment]
Can either one of you officially approve this to unblock the merge?

**Comment by youkaichao** [comment]
You need to also add ray backend test.

**Review by youkaichao — APPROVED** [review]
LGTM, cc @ywang96

**Pull request merged**